No-code guide to state machines and statecharts (#3516)
statelyai/xstate: docs/about/glossary.md

# Glossary

This glossary is a guide to the most common terms in statecharts and state machines.

## State machines

> A state machine is a model showing how you transition from state to state in a process when events occur.

Reviewer: ```suggestion
A state machine is an application model that ensures it will only be in one state at a time.
```

I stumbled a bit on the description of a state machine. Not sure my suggestion is much better though 😆

Reviewer 2: Discussed on Discord:

```suggestion
A state machine is a model that describes how the state of a process transitions to another state when an event occurs. State machines make building reliable software easier because they prevent impossible states and undesired transitions.
```

> ## Statecharts

Reviewer: ```suggestion
## State charts
```

> Statecharts are a visual language used to visualize state machines. You may have used similar diagrams in the past to design user flows, plan databases or map app architecture. Statecharts are another way of using boxes and arrows to represent flows. These flows are also executable code you can use to control the logic in your applications.

Reviewer: ```suggestion
State charts are a visual language used to visualize state machines. You may have used similar diagrams in the past to design user flows, plan databases or map app architecture. Statecharts are another way of using boxes and arrows to represent flows. These flows are also executable code you can use to control the logic in your applications.
```

> Statecharts are a visual language used to visualize state machines. You may have used similar diagrams in the past to design user flows, plan databases or map app architecture. Statecharts are another way of using boxes and arrows to represent flows. These flows are also executable code you can use to control the logic in your applications.

Reviewer 2: ```suggestion
Statecharts are a visual language used to express complex state machines in a simpler way. You may have used similar diagrams in the past to design user flows, plan data transformations or map app logic. Statecharts are another way of using boxes and arrows to represent these kinds of flows. These flows are also executable code you can use to control the logic directly in your application code.
```

![basic statechart with an initial state transitioning through an event to another state, then transitioning through another event to a final state.](./basic-statechart.png)

## States

A state describes the status of the machine. A state can be as simple as _on_ and _off_. These states are finite; the machine can only move through the pre-defined states. And machines can only be in one state at a time, unless it is a [parent state](#parent-and-child-states) or a [parallel state](#parallel-states).

### Initial state

> When a state machine starts, it enters the **initial state** first. The initial state is represented by the <img alt='filled circle with an arrow pointing from the circle to the initial state' src='./initial-state-icon.png' height='24' width='24' style='display: inline; margin-bottom: -5px' /> icon. A machine can only have one initial state.

Reviewer 2: ```suggestion
When a state machine starts, it enters the **initial state** first. The initial state is represented by the <img alt='filled circle with an arrow pointing from the circle to the initial state' src='./initial-state-icon.png' height='24' width='24' style='display: inline; margin-bottom: -5px' /> icon. A machine can only have one top-level initial state.
```

<!-- What is a typical event -->

### Guarded transitions

A guard is a condition that the machine checks when it goes through an event. If the condition is true, the machine follows the transition to the next state. If the condition is false, the machine follows the rest of the conditions to the next state. Any transition can be a guarded transition.

<!-- What you might use a guard for -->

### Eventless transitions

Eventless transitions are transitions without events. These transitions **always** run when the machine goes through the event; no event is necessary to trigger the transition. Eventless transitions are labeled “always.”

<!-- What you might use an eventless transition for -->

### Wildcard transitions

Wildcard transitions are triggered by any event not already handled by the current state. Wildcard transitions are represented by a \*.

Wildcard transitions are useful for logging untracked events and reducing code duplication.

<!-- What you might use a wildcard transition for -->

<!-- #### Partial wildcard transitions -->

<!-- Will be in v5 -->

<!-- ### Raised events -->

<!-- Will be in v5 -->

## Actors, actions and invocations

> Statecharts are executable code. When you run a statechart, it becomes an actor; a running process that can receive messages, send messages and change its behavior based on the messages it receives, causing effects outside the machine.

Reviewer 2: ```suggestion
A statechart is an executable model of an actor. When you run a statechart, it becomes an actor; a running process that can receive messages, send messages and change its behavior based on the messages it receives, which can cause effects outside of the actor.
```
